### PR TITLE
gdcm: conan v2 support

### PR DESCRIPTION
**Pull request by Nekto89**
Specify library name and version:  **gdcm/3.0.9**

conan v2 support

Fix compilation with gcc11 and bump dependencies.

Patch was taken from https://github.com/malaterre/GDCM/commit/4404b770be337bd0d5d3c2289abfd34426433db2

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by Nekto89** [comment]
> @Nekto89 Thank you for your contribution. The Conan v2 linter found some outdated imports. Could you please update them? As reference, you can use https://github.com/conan-io/conan-center-index/blob/master/docs/v2_linter.md

I've tried but I'm having some problems with expat. GDCM uses "EXPAT_INCLUDE_DIRS" variable that should be set from generated "FindEXPAT.cmake". Inside it I see "set(EXPAT_INCLUDE_DIRS ${expat_INCLUDE_DIRS_RELEASE} )". But after call to find_package only expat_INCLUDE_DIRS_RELEASE is present.

```
find_package(EXPAT REQUIRED)
message("INCLUDE_DIRS_1 ${EXPAT_INCLUDE_DIRS}")
message("INCLUDE_DIRS_2 ${expat_INCLUDE_DIRS_RELEASE}")
```
```
1>-- Conan: Target declared 'openjp2'
1>-- Conan: Target declared 'ZLIB::ZLIB'
1>-- Conan: Target declared 'expat::expat'
1>INCLUDE_DIRS_1
1>INCLUDE_DIRS_2 ....../.conan/data/expat/2.4.9/_/_/package/8e589e066a19f700666be77ed94ff8e8bc88ee10/include
```

Does this mean that additional patching of GDCM is required for using CMakeToolchain/CMakeDeps?

**Comment by Nekto89** [comment]
changing "find_package(EXPAT REQUIRED)" to "find_package(EXPAT REQUIRED MODULE)" fixes the issue. Is this intended conan behavior?

**Comment by Nekto89** [comment]
Can I use `self.dependencies["expat"].cpp_info.set_property("cmake_find_mode", "module")`? Will it also affect other recipes?

**Comment by uilianries** [comment]
> Can I use `self.dependencies["expat"].cpp_info.set_property("cmake_find_mode", "module")`? Will it also affect other recipes?

Nope. `self.dependencies` is read-only. It you need to change anything, you have to fix expat, or patch gdcm.

File: recipes/gdcm/all/conanfile.py [hunk 1]
```diff
@@ -25,101 +30,89 @@ class GDCMConan(ConanFile):
...
+        tc.cache_variables["GDCM_USE_SYSTEM_OPENJPEG"] = True
+        tc.cache_variables["GDCM_USE_SYSTEM_ZLIB"] = True
+        if not self.settings.compiler.cppstd:
+            tc.blocks["cppstd"].values = { "cppstd": self._minimum_cpp_standard, "cppstd_extensions": "OFF" }
+        tc.generate()
+        deps = CMakeDeps(self)
+        previous_value = self.dependencies["expat"].cpp_info.get_property("cmake_find_mode")
+        self.dependencies["expat"].cpp_info.set_property("cmake_find_mode", "module")
+        deps.generate()
+        self.dependencies["expat"].cpp_info.set_property("cmake_find_mode", previous_value)
```
Comment (uilianries) [inline recipes/gdcm/all/conanfile.py:84]: `self.dependencies` is read-only: https://docs.conan.io/en/latest/reference/conanfile/dependencies.html#dependencies-interface

**Review by uilianries — CHANGES_REQUESTED** [review]

File: recipes/gdcm/all/conanfile.py (hunk 1 above)
Comment (Nekto89) [reply recipes/gdcm/all/conanfile.py]: fixed

File: recipes/gdcm/all/conanfile.py [hunk 2]
```diff
@@ -25,101 +30,86 @@ class GDCMConan(ConanFile):
...
-        self._cmake.definitions["GDCM_USE_SYSTEM_ZLIB"] = True
-        if not tools.valid_min_cppstd(self, 11):
-            self._cmake.definitions["CMAKE_CXX_STANDARD"] = 11
-
-        self._cmake.configure(build_folder=self._build_subfolder)
-        return self._cmake
+        tc.cache_variables["GDCM_USE_SYSTEM_EXPAT"] = True
+        tc.cache_variables["GDCM_USE_SYSTEM_OPENJPEG"] = True
+        tc.cache_variables["GDCM_USE_SYSTEM_ZLIB"] = True
+        if not self.settings.compiler.cppstd:
```
Comment (Nekto89) [inline recipes/gdcm/all/conanfile.py:77]: @uilianries is it possible to have same package id with compiler.cppstd=11 and without it?
I've tried it but got error
```
def package_id(self):
        if not self.info.settings.compiler.cppstd:
            self.info.settings.compiler.cppstd = self._minimum_cpp_standard
```
`gdcm/3.0.9: Invalid ID: The compiler.cppstd is not defined for this configuration`

Comment (Nekto89) [reply recipes/gdcm/all/conanfile.py:77]: hm, it seems that "11" can't be even used with conan. Tried compatible packages
```
def package_id(self):
        if self.settings.compiler.cppstd == self._minimum_cpp_standard:
            compatible_pkg = self.info.clone()
            del compatible_pkg.compiler.cppstd
            self.compatible_packages.append(compatible_pkg)
```
ConanException: Invalid setting '11' is not a valid 'settings.compiler.cppstd' value.
Possible values are ['None', '14', '17', '20', '23']

Comment (uilianries) [reply recipes/gdcm/all/conanfile.py:77]: @Nekto89 

> is it possible to have same package id with compiler.cppstd=11 and without it?

No. However, you can manage the package ID: https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html#defining-a-custom-package-id

However, customizing at this level is **forbidden** in CCI.

Comment (uilianries) [reply recipes/gdcm/all/conanfile.py:77]: > hm, it seems that "11" can't be even used with conan. Tried compatible packages

You customized your settings.yml for sure. `11` is always present there by default.

You can remove `settings.yml` and run conan create again, then Conan will generate a new settings.yml for you

File: recipes/gdcm/all/conanfile.py [hunk 3]
```diff
@@ -25,101 +30,86 @@ class GDCMConan(ConanFile):
...
+        get(self, **self.conan_data["sources"][self.version],
+                  destination=self.source_folder, strip_root=True)
 
-        self._cmake = CMake(self)
-        self._cmake.definitions["GDCM_BUILD_DOCBOOK_MANPAGES"] = False
-        self._cmake.definitions["GDCM_BUILD_SHARED_LIBS"] = self.options.shared
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.cache_variables["GDCM_BUILD_DOCBOOK_MANPAGES"] = False
+        tc.cache_variables["GDCM_BUILD_SHARED_LIBS"] = self.info.options.shared
```
Comment (uilianries) [inline recipes/gdcm/all/conanfile.py:72]:
```suggestion
        tc.cache_variables["GDCM_BUILD_SHARED_LIBS"] = self.options.shared
```

`self.info.options` is for `validate()` method

Comment (Nekto89) [reply recipes/gdcm/all/conanfile.py:72]: this leads to 
```
tc.generate()
        TypeError: Object of type PackageOption is not JSON serializable
```

Comment (Nekto89) [reply recipes/gdcm/all/conanfile.py:72]:
```suggestion
        tc.cache_variables["GDCM_BUILD_SHARED_LIBS"] = bool(self.options.shared)
```

File: recipes/gdcm/all/conanfile.py (hunk 2 above)
Comment (Nekto89) [reply recipes/gdcm/all/conanfile.py:77]: it seems that Visual Studio doesn't have C++11 as separate value for /std: option. Having CMAKE_CXX_STANDARD 11 leads to the same result as not having CMAKE_CXX_STANDARD at all.

File: recipes/gdcm/all/test_package/conanfile.py [hunk 4]
```diff
@@ -1,21 +1,30 @@
...
-            bin_path = os.path.join("bin", "test_package")
+        if can_run(self):
+            bin_path = os.path.join(self.cpp.build.bindirs[0], "test_package")
             input_file = os.path.join(self.source_folder, "DCMTK_JPEGExt_12Bits.dcm")
             test_dir = "test_dir"
-            tools.mkdir(test_dir)
+            mkdir(self, test_dir)
             output_file = os.path.join(test_dir, "output.dcm")
-            self.run([bin_path, input_file, output_file], run_environment=True)
+            self.run(f"\"{bin_path}\" \"{input_file}\" \"{output_file}\"", env="conanrun")
```
Comment (uilianries) [inline recipes/gdcm/all/test_package/conanfile.py:30]:
```suggestion
            self.run(f"'{bin_path}' '{input_file}' '{output_file}'", env="conanrun")
```

**Review by uilianries — DISMISSED** [review]

**Review by uilianries — DISMISSED** [review]

File: recipes/gdcm/all/conanfile.py [hunk 5]
```diff
@@ -25,101 +30,86 @@ class GDCMConan(ConanFile):
...
+        get(self, **self.conan_data["sources"][self.version],
+                  destination=self.source_folder, strip_root=True)
 
-        self._cmake = CMake(self)
-        self._cmake.definitions["GDCM_BUILD_DOCBOOK_MANPAGES"] = False
-        self._cmake.definitions["GDCM_BUILD_SHARED_LIBS"] = self.options.shared
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.cache_variables["GDCM_BUILD_DOCBOOK_MANPAGES"] = False
+        tc.cache_variables["GDCM_BUILD_SHARED_LIBS"] = bool(self.options.shared)
```
Comment (prince-chrismc) [inline recipes/gdcm/all/conanfile.py:72]:
```suggestion
        tc.variables["GDCM_BUILD_DOCBOOK_MANPAGES"] = False
        tc.variables["GDCM_BUILD_SHARED_LIBS"] = bool(self.options.shared)
```

File: recipes/gdcm/all/conanfile.py [hunk 6]
```diff
@@ -25,101 +30,86 @@ class GDCMConan(ConanFile):
...
-        if not tools.valid_min_cppstd(self, 11):
-            self._cmake.definitions["CMAKE_CXX_STANDARD"] = 11
-
-        self._cmake.configure(build_folder=self._build_subfolder)
-        return self._cmake
+        tc.cache_variables["GDCM_USE_SYSTEM_EXPAT"] = True
+        tc.cache_variables["GDCM_USE_SYSTEM_OPENJPEG"] = True
+        tc.cache_variables["GDCM_USE_SYSTEM_ZLIB"] = True
+        if not self.settings.compiler.cppstd:
+            tc.cache_variables["CMAKE_CXX_STANDARD"] = self._minimum_cpp_standard
```
Comment (prince-chrismc) [inline recipes/gdcm/all/conanfile.py:78]:
```suggestion
        tc.variables["GDCM_USE_SYSTEM_EXPAT"] = True
        tc.variables["GDCM_USE_SYSTEM_OPENJPEG"] = True
        tc.variables["GDCM_USE_SYSTEM_ZLIB"] = True
        if not self.settings.compiler.cppstd:
            tc.variables["CMAKE_CXX_STANDARD"] = self._minimum_cpp_standard
```

Comment (prince-chrismc) [inline recipes/gdcm/all/conanfile.py:78]: We almost never do this in CCI... we typically raise for the compiler version when the default is not high enough 😕 

Not sure what the right call is for this... just thinking out loud

**Review by prince-chrismc — COMMENTED** [review]
cache variables should be used sparingly. the default variable is more then enough

**Comment by Nekto89** [comment]
> cache variables should be used sparingly. the default variable is more then enough

Maybe this information should be added to package template?
[pack](https://github.com/conan-io/conan-center-index/blob/master/docs/package_templates/cmake_package/all/conanfile.py)

File: recipes/gdcm/all/conanfile.py (hunk 6 above)
Comment (uilianries) [reply recipes/gdcm/all/conanfile.py]: some projects use C++11 features, but they are not configured in CMake. I believe the author runs some newer version of GCC, which runs C++11 by default, but forget about older versions.

**Review by jwillikers — APPROVED** [review]

**Comment by prince-chrismc** [comment]
> Maybe this information should be added to package template?

https://github.com/conan-io/conan-center-index/pull/13385 👍

**Review by prince-chrismc — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Pull request merged**